cohortextractor/validate_dummy_data.py

```diff
@@ -0,0 +1,114 @@
+from datetime import datetime
+import pandas as pd
+
+
+def validate_dummy_data(study_definition, dummy_data_file):
+    """Validate that dummy data provided by user matches expected structure and format.
+
+    Raises DummyDataValidationError if dummy data is not valid.
+    """
+
+    df = read_into_dataframe(dummy_data_file)
+
+    covariate_definitions = study_definition.covariate_definitions
+
+    for col_name in df:
+        if col_name == "patient_id":
+            # patient_id is not present in covariate_definitions
+            continue
+
+        if col_name not in covariate_definitions:
+            raise DummyDataValidationError(
+                f"Unexpected column {col_name} in dummy data"
+            )
+
+    for (col_name, (_, query_args)) in covariate_definitions.items():
+        if col_name == "population":
+            # Ignore the population definition, since it is not used as a column in the
+            # output
+            continue
+
+        if query_args["hidden"]:
+            # Ignore hidden covariates
+            continue
+
+        if col_name not in df:
+            raise DummyDataValidationError(
+                f"Column {col_name} is missing from dummy data"
+            )
+
+        validator = get_validator(col_name, query_args)
+
+        for ix, value in enumerate(df[col_name]):
+            if pd.isna(value) or value == "":
+                # Ignore null or missing values
+                continue
+
+            try:
+                validator(value)
+            except ValueError:
+                raise DummyDataValidationError(
+                    f"Invalid value `{value}` for {col_name} in row {ix + 2}"
+                )
+
+
+def read_into_dataframe(path):
+    """Read data from path into a Pandas DataFrame."""
+
+    if path.suffixes in [[".csv"], [".csv", ".gz"]]:
+        return pd.read_csv(path)
+    elif path.suffixes in [[".dta"], [".dta", ".gz"]]:
+        return pd.read_stata(path)
+    elif path.suffix == ".feather":
+        return pd.read_feather(path)
+    else:
+        assert False, path
+
+
+def get_validator(col_name, query_args):
```

Comment:
These validators are, I think, specific to CSV files and won't work with binary formats. At least the date ones certainly are and the `int`, `str`, `float` checks will test that columns can be cast to the right type but not that they're the right type initially.

Perhaps this function could take an extra argument which is the file format so it can return the appropriate validator. In practice I hope we'd only need two sets: one for CSV and another for all other binary formats.